src/Shared/NodeEndpointOutOfProcBase.cs

```diff
@@ -378,53 +371,67 @@ private void PacketPumpProc()
                     try
                     {
-                        long handshake = localReadPipe.ReadLongForHandshake(/* reject these leads */ new byte[] { 0x5F, 0x60 }, 0xFF /* this will disconnect the host; it expects leading 00 or F5 or 06 */
+                        int index = 1;
+                        foreach (int part in handshake.RetrieveHandshakeComponents())
+                        {
+
+                            int handshakePart = localReadPipe.ReadIntForHandshake(index == 1 ? 0x01 : 0x00 /* this will disconnect a < 4.5 host; it expects leading 00 or F5 or 06 */
 #if NETCOREAPP2_1 || MONO
                             , ClientConnectTimeout /* wait a long time for the handshake from this side */
 #endif
                             );
 
-#if FEATURE_SECURITY_PERMISSIONS
-                        WindowsIdentity currentIdentity = WindowsIdentity.GetCurrent();
-#endif
-
-                        if (handshake != GetHostHandshake())
-                        {
-                            CommunicationsUtilities.Trace("Handshake failed. Received {0} from host not {1}. Probably the host is a different MSBuild build.", handshake, GetHostHandshake());
-                            localPipeServer.Disconnect();
-                            continue;
+                            if (handshakePart != part)
+                            {
+                                CommunicationsUtilities.Trace("Handshake failed. Received {0} from host not {1}. Probably the host is a different MSBuild build.", handshakePart, part);
+                                localWritePipe.WriteIntForHandshake(index);
+                                gotValidConnection = false;
+                                break;
+                            }
+                            index++;
                         }
 
-#if FEATURE_SECURITY_PERMISSIONS
-                        // We will only talk to a host that was started by the same user as us.  Even though the pipe access is set to only allow this user, we want to ensure they
-                        // haven't attempted to change those permissions out from under us.  This ensures that the only way they can truly gain access is to be impersonating the
-                        // user we were started by.
-                        WindowsIdentity clientIdentity = null;
-                        localPipeServer.RunAsClient(delegate () { clientIdentity = WindowsIdentity.GetCurrent(true); });
+                        CommunicationsUtilities.Trace("Writing accept code to parent");
+                        localWritePipe.WriteIntForHandshake(0x39393939);
 
-                        if (clientIdentity == null || !String.Equals(clientIdentity.Name, currentIdentity.Name, StringComparison.OrdinalIgnoreCase))
+                        // To ensure that our handshake and theirs have the same number of bytes, send and receive a magic number indicating EOS. This has to be different from the
```

Comment:
Thinking out loud: I wonder if it would be worth having "SendEOS" or "sendAccept" style functions that simply writes the magic number, but the name of the function gives it context instead of commenting in multiple places.

Reply:
That's clean and pretty easy. Thanks!